File: src/main/scala/org/apache/spark/sql/Trigger.scala
package org.apache.spark.sql

import org.apache.spark.sql.execution.streaming.*

/** Compatibility facade for the historical root-package Trigger.
  *
  * Prefer [[org.apache.spark.sql.streaming.Trigger]], which matches upstream Spark Connect.
  */
type Trigger = streaming.Trigger

object Trigger:
  type ProcessingTime = ProcessingTimeTrigger

  object ProcessingTime:
    def apply(intervalMs: Long): ProcessingTimeTrigger =
      ProcessingTimeTrigger.create(intervalMs, java.util.concurrent.TimeUnit.MILLISECONDS)

    def apply(
        interval: Long,
        timeUnit: java.util.concurrent.TimeUnit
    ): ProcessingTimeTrigger =
      ProcessingTimeTrigger.create(interval, timeUnit)

    def apply(interval: scala.concurrent.duration.Duration): ProcessingTimeTrigger =
      ProcessingTimeTrigger(interval)

    def apply(interval: String): ProcessingTimeTrigger =
      ProcessingTimeTrigger(interval)

    def unapply(trigger: ProcessingTimeTrigger): Option[Long] =
      Some(trigger.intervalMs)

  def processingTime(intervalMs: Long): ProcessingTimeTrigger =
    ProcessingTime(intervalMs)

  def processingTime(
      interval: Long,
      timeUnit: java.util.concurrent.TimeUnit
  ): ProcessingTimeTrigger =
    ProcessingTime(interval, timeUnit)

  def processingTime(
      interval: scala.concurrent.duration.Duration
  ): ProcessingTimeTrigger =
    ProcessingTime(interval)

  def processingTime(interval: String): ProcessingTimeTrigger =
    ProcessingTime(interval)

  val AvailableNow: streaming.Trigger =
    streaming.Trigger.AvailableNow()

  val Once: streaming.Trigger =
    streaming.Trigger.Once()

  type Continuous = ContinuousTrigger

  object Continuous:
    def apply(intervalMs: Long): ContinuousTrigger =
      ContinuousTrigger(intervalMs)

    def apply(
        interval: Long,
        timeUnit: java.util.concurrent.TimeUnit
    ): ContinuousTrigger =
      ContinuousTrigger.create(interval, timeUnit)

    def apply(interval: scala.concurrent.duration.Duration): ContinuousTrigger =
      ContinuousTrigger(interval)

    def apply(interval: String): ContinuousTrigger =
      ContinuousTrigger(interval)

    def unapply(trigger: ContinuousTrigger): Option[Long] =
      Some(trigger.intervalMs)

  def continuous(intervalMs: Long): ContinuousTrigger =
    Continuous(intervalMs)

  def continuous(
      interval: Long,
      timeUnit: java.util.concurrent.TimeUnit
  ): ContinuousTrigger =
    Continuous(interval, timeUnit)

  def continuous(interval: scala.concurrent.duration.Duration): ContinuousTrigger =
    Continuous(interval)

  def continuous(interval: String): ContinuousTrigger =
    Continuous(interval)

  type RealTime = RealTimeTrigger

  object RealTime:
    def apply(): RealTimeTrigger =
      RealTimeTrigger()

    def apply(batchDurationMs: Long): RealTimeTrigger =
      RealTimeTrigger(batchDurationMs)

    def apply(
        batchDuration: Long,
        timeUnit: java.util.concurrent.TimeUnit
    ): RealTimeTrigger =
      RealTimeTrigger.create(batchDuration, timeUnit)

    def apply(batchDuration: scala.concurrent.duration.Duration): RealTimeTrigger =
      RealTimeTrigger(batchDuration)

    def apply(batchDuration: String): RealTimeTrigger =
      RealTimeTrigger(batchDuration)

    def unapply(trigger: RealTimeTrigger): Option[Long] =
      Some(trigger.batchDurationMs)

  def realTime(): RealTimeTrigger =
    RealTime()

  def realTime(batchDurationMs: Long): RealTimeTrigger =
    RealTime(batchDurationMs)

  def realTime(
      batchDuration: Long,
      timeUnit: java.util.concurrent.TimeUnit
  ): RealTimeTrigger =
    RealTime(batchDuration, timeUnit)

  def realTime(batchDuration: scala.concurrent.duration.Duration): RealTimeTrigger =
    RealTime(batchDuration)

  def realTime(batchDuration: String): RealTimeTrigger =
    RealTime(batchDuration)

File: src/main/java/org/apache/spark/sql/streaming/Trigger.java
package org.apache.spark.sql.streaming;

import java.util.concurrent.TimeUnit;

import scala.concurrent.duration.Duration;

import org.apache.spark.annotation.Evolving;
import org.apache.spark.annotation.Experimental;
import org.apache.spark.sql.execution.streaming.AvailableNowTrigger$;
import org.apache.spark.sql.execution.streaming.ContinuousTrigger$;
import org.apache.spark.sql.execution.streaming.OneTimeTrigger$;
import org.apache.spark.sql.execution.streaming.ProcessingTimeTrigger$;
import org.apache.spark.sql.execution.streaming.RealTimeTrigger$;

/** Policy used to indicate how often results should be produced by a streaming query. */
@Evolving
public class Trigger {
  public static Trigger ProcessingTime(long intervalMs) {
    return ProcessingTimeTrigger$.MODULE$.create(intervalMs, TimeUnit.MILLISECONDS);
  }

  public static Trigger ProcessingTime(long interval, TimeUnit timeUnit) {
    return ProcessingTimeTrigger$.MODULE$.create(interval, timeUnit);
  }

  public static Trigger ProcessingTime(Duration interval) {
    return ProcessingTimeTrigger$.MODULE$.apply(interval);
  }

  public static Trigger ProcessingTime(String interval) {
    return ProcessingTimeTrigger$.MODULE$.apply(interval);
  }

  @Deprecated(since = "3.4.0")
  public static Trigger Once() {
    return OneTimeTrigger$.MODULE$;
  }

  public static Trigger AvailableNow() {
    return AvailableNowTrigger$.MODULE$;
  }

  public static Trigger Continuous(long intervalMs) {
    return ContinuousTrigger$.MODULE$.apply(intervalMs);
  }

  public static Trigger Continuous(long interval, TimeUnit timeUnit) {
    return ContinuousTrigger$.MODULE$.create(interval, timeUnit);
  }

  public static Trigger Continuous(Duration interval) {
    return ContinuousTrigger$.MODULE$.apply(interval);
  }

  public static Trigger Continuous(String interval) {
    return ContinuousTrigger$.MODULE$.apply(interval);
  }

  @Experimental
  public static Trigger RealTime(long batchDurationMs) {
    return RealTimeTrigger$.MODULE$.apply(batchDurationMs);
  }

  @Experimental
  public static Trigger RealTime(long batchDuration, TimeUnit timeUnit) {
    return RealTimeTrigger$.MODULE$.create(batchDuration, timeUnit);
  }

  @Experimental
  public static Trigger RealTime(Duration batchDuration) {
    return RealTimeTrigger$.MODULE$.apply(batchDuration);
  }

  @Experimental
  public static Trigger RealTime(String batchDuration) {
    return RealTimeTrigger$.MODULE$.apply(batchDuration);
  }

  @Experimental
  public static Trigger RealTime() {
    return RealTimeTrigger$.MODULE$.apply();
  }
}
